src/analysis/ReachingDefinitions/Srg/IntervalMap.h

```diff
@@ -0,0 +1,214 @@
+#ifndef _DG_INTERVALSET_H
+#define _DG_INTERVALSET_H
+
+#include "analysis/Offset.h"
+#include "analysis/ReachingDefinitions/RDMap.h"
+
+namespace dg {
+namespace analysis {
+namespace rd {
+namespace srg {
+namespace detail {
+
+/**
+ * represents a both-side closed interval of values
+ */
+class Interval {
+    using T = Offset;
+
+    T start;
+    T len;
+
+    static inline T min(T a, T b) {
+        return a < b ? a : b;
+    }
+
+    static inline T max(T a, T b) {
+        return a > b ? a : b;
+    }
+
+public:
+    Interval(T start, T len): start(start), len(len) {}
+
+    bool isUnknown() const {
+        return start.isUnknown() || len == 0;
+    }
+
+    bool overlaps(const Interval& other) const {
+        if (isUnknown() || other.isUnknown())
+            return false;
+        return intervalsOverlap(start.offset, len.offset, other.start.offset, other.len.offset);
+    }
+
+    bool isSubsetOf(const Interval& other) const {
```

Comment:
I'm not sure, but what if `isUnknown()` is true for one `this` or `other`?

Reply:
At the moment, `IntervalMap` is used in `MarkerSRGBuilderFS`, where all intervals that return `true` from `isUnknown()` function are stretched to full size(offset=0, len=`~((uint64_t) 0)`) before being passed to `IntervalMap`. Thinking about it, I should probably add assertions to check that manipulated intervals all have known offsets or maybe move the stretching function into `IntervalMap`.